src/log.c

```diff
@@ -36,70 +36,297 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
 
+#include <rdma/fi_errno.h>
 #include "fi.h"
 #include "fi_log.h"
 
+#define ARRAY_SIZE(x) (sizeof((x)) / sizeof((x)[0]))
+
+static struct log_providers provider_list;
+
+static const char *subsystems[] = {
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[CM] = "cm",
+	[EP] = "ep",
+	[AV] = "av",
+	[CQ] = "cq",
+	[EQ] = "eq",
+	[MR] = "mr"
+};
+
+static const char *log_levels[] = {
+	[LOG_WARN] = "warn",
+	[LOG_TRACE] = "trace",
+	[LOG_INFO] = "info"
+};
+
+/*
+ * By default support logging from all providers and subsystems. This will
+ * change if the relevant environment variables are set.
+ */
+uint64_t log_mask = PROV_MASK | SUBSYS_MASK;
+
+typedef void found_func_t(char *ptr);
+
 /* General implementation note: these functions currently use multiple fprintfs
  * in a row, which can render in an ugly fashion for multithreaded code and for
  * some mpirun implementations.  If this bugs anyone enough then we can convert
  * them to snprintf to build up the printout in a single buffer.
```

Comment:
@bturrubiates @jsquyres Again throwing my two cents in: for EXS's logging, I use `flockfile` and `funlockfile` around multiple `printf`s, which works since the file locks are defined by POSIX to be reentrant.  Doing essentially the following seems to work for EXS, and the overhead hasn't cost me enough performance to make the debug output useless:

```patch
diff --git a/src/log.c b/src/log.c
index 64c79f8..6f76e58 100644
--- a/src/log.c
+++ b/src/log.c
@@ -68,6 +68,8 @@ void fi_warn_impl(const char *prov, const char *fmt, ...)
 {
        va_list vargs;
 
+       pthread_cleanup_push((void (*)(void *))funlockfile, stderr);
+       flockfile(stderr);
        if (prov != NULL)
                fprintf(stderr, "%s:%s: ", PACKAGE, prov);
        else
@@ -75,6 +77,7 @@ void fi_warn_impl(const char *prov, const char *fmt, ...)
        va_start(vargs, fmt);
        vfprintf(stderr, fmt, vargs);
        va_end(vargs);
+       pthread_cleanup_pop(1);
 }
```

Note that you really should do something about thread cancellation if you grab file locks, since printf calls into cancellation points.  OTOH, that adds a probably not insignificant amount of overhead, so maybe we just say that you should not cancel a thread when it might be in an OFI function call.